core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -176,10 +179,28 @@ class LogSegment private[log] (val log: FileRecords,
       throw new LogSegmentOffsetOverflowException(this, offset)
   }
 
+  /**
+   * Validate that offsets of records are monotonically increasing and are in [offsetLowerBound, offsetUpperBound] range.
+   *
+   * @param records The log records to validate
+   * @param offsetLowerBound The lower bound of offsets of all records (inclusive)
+   * @param offsetUpperBound The upper bound of offsets of all records (inclusive)
+   */
+  private def ensureOffsetsInRangeForRecords(records: Records, offsetLowerBound: Long, offsetUpperBound: Long): Unit = {
+    var offsetLowerBoundForBatch = offsetLowerBound
+    records.batches.forEach { batch =>
+      if (batch.baseOffset < offsetLowerBoundForBatch || batch.lastOffset > offsetUpperBound || batch.baseOffset > batch.lastOffset) {
```

Comment:
Was the issue reported issue in the jira on the leader or the follower?  If it's on the leader, currently, each leader append only has one batch. I am wondering if the validation should be done inside LogValidator.validateMessagesAndAssignOffsets.